src/ast/parse.rs

```diff
@@ -0,0 +1,895 @@
+//! Parse a string into egglog.
+
+use crate::*;
+use ordered_float::OrderedFloat;
+
+pub fn parse_program(filename: Option<String>, input: &str) -> Result<Vec<Command>, ParseError> {
+    let (out, _span, rest) = program(&Context::new(filename, input))?;
+    assert!(rest.is_at_end(), "did not parse entire program");
+    Ok(out)
+}
+
+// currently only used for testing, but no reason it couldn't be used elsewhere later
+pub fn parse_expr(filename: Option<String>, input: &str) -> Result<Expr, ParseError> {
+    let (out, _span, rest) = expr(&Context::new(filename, input))?;
+    assert!(rest.is_at_end(), "did not parse entire expression");
+    Ok(out)
+}
+
+/// A [`Span`] contains the file name and a pair of offsets representing the start and the end.
+#[derive(Clone, PartialEq, Eq, Hash)]
+pub struct Span(Arc<SrcFile>, usize, usize);
+
+lazy_static::lazy_static! {
+    pub static ref DUMMY_SPAN: Span = Span(Arc::new(SrcFile {name: None, contents: String::new()}), 0, 0);
+}
+
+impl Span {
+    pub fn string(&self) -> &str {
+        &self.0.contents[self.1..self.2]
+    }
+}
+
+#[derive(Debug, PartialEq, Eq, Hash)]
+struct SrcFile {
+    name: Option<String>,
+    contents: String,
+}
+
+struct Location {
+    line: usize,
+    col: usize,
+}
+
+impl SrcFile {
+    pub fn get_location(&self, offset: usize) -> Location {
+        let mut line = 1;
+        let mut col = 1;
+        for (i, c) in self.contents.char_indices() {
```

Comment:
optional: 

Should we build an index when constructing `SrcFile` that keeps a btree map mapping from position of a newline to an increasing count of newlines?

That way, we could use the `upper_bound` method to essentially seek to the line in question and compute this all in log(n) time?.

Maybe it's not worth it though. I don't think it'd be that much code though.

Reply:
`get_location` is allowed to be slow because it only ever gets called twice-ish, and only on the error path.